reasoner/planner/AnswerCountEstimator.java

```diff
@@ -0,0 +1,486 @@
+    private final Set<ResolvableConjunction> cycleHeads;
+    private final Set<ResolvableConjunction> toReset;
+    private final AnswerCountModel answerCountModel;
+
+    public AnswerCountEstimator(LogicManager logicMgr) {
+        this.logicMgr = logicMgr;
+        this.answerCountModel = new AnswerCountModel(this);
+        this.estimators = new HashMap<>();
+        this.initializationStack = new ArrayList<>();
+        this.cycleHeads = new HashSet<>();
+        this.toReset = new HashSet<>();
+    }
+
+    public long estimateAllAnswers(ResolvableConjunction conjunction) {
+        registerConjunctionAndInitialize(conjunction);
+        return estimators.get(conjunction).estimateAllAnswers();
+    }
+
+    public long estimateAnswers(ResolvableConjunction conjunction, Set<Variable> variableFilter) {
+        return estimateAnswers(conjunction, variableFilter, logicMgr.compile(conjunction));
+    }
+
+    public long estimateAnswers(ResolvableConjunction conjunction, Set<Variable> variableFilter, Set<Resolvable<?>> includedResolvables) {
+        registerConjunctionAndInitialize(conjunction);
+        return estimators.get(conjunction).estimateAnswers(variableFilter, includedResolvables);
+    }
+
+    private void registerConjunctionAndInitialize(ResolvableConjunction conjunction) {
+        if (!estimators.containsKey(conjunction)) {
+            estimators.put(conjunction, new ConjunctionAnswerCountEstimator(this, conjunction));
+        }
+
+        if (estimators.get(conjunction).mayInitialize()) {
+            initializationStack.add(conjunction);
+            estimators.get(conjunction).initialize();
+            assert initializationStack.get(initializationStack.size() - 1) == conjunction;
+            initializationStack.remove(initializationStack.size() - 1);
+
+            if (cycleHeads.size() == 1 && cycleHeads.contains(conjunction)) {
+                // Reset all cyclic paths
+                for (ResolvableConjunction conjunctionToReset : toReset) {
+                    estimators.get(conjunctionToReset).reset();
+                }
+            }
+        }
+    }
+
+    private void reportInitializationCycle(ResolvableConjunction conjunction) {
+        cycleHeads.add(conjunction);
+        for (int i = initializationStack.size() - 1; initializationStack.get(i) != conjunction; i--) {
+            toReset.add(initializationStack.get(i));
+        }
+    }
+
+    private static class ConjunctionAnswerCountEstimator {
+
+        private final ResolvableConjunction conjunction;
+        private final AnswerCountEstimator answerCountEstimator;
+        private final LogicManager logicMgr;
+        private final AnswerCountModel answerCountModel;
+        private Map<Variable, LocalEstimate> unaryEstimateCover;
+        private Map<Resolvable<?>, List<LocalEstimate>> retrievableEstimates;
+        private Map<Resolvable<?>, LocalEstimate> inferrableEstimates;
+        private Map<Resolvable<?>, LocalEstimate> unaryEstimates;
```

Comment:
a `Map` can usually be final, and initialised in the constructor. `reset()` can then `clear()` the map.

I believe you're making them nullable also to know when to recompute these values - however I expect that this signal should be derivable from the `Status` flag!

Reply:
I agree that the status flag is the ground-truth for deciding whether to (re-)compute the fields.
But I would rather not initialize them to an empty set, since the final result can also be an empty set. Isn't it better to have a loud crash when it's accessed before it's been computed, than to have a silent failure?